state/raftlease/target.go

```diff
@@ -174,33 +176,110 @@ func (t *notifyTarget) Claimed(key lease.Key, holder string) error {
 	return errors.Annotatef(err, "%q for %q in db", docId, holder)
 }
 
-// Expired is part of raftlease.NotifyTarget.
-func (t *notifyTarget) Expired(key lease.Key) error {
+type leaseDoc struct {
+	Key    lease.Key
+	DocId  string
+	Holder string
+}
+
+// Expiries is part of raftlease.NotifyTarget.
+func (t *notifyTarget) Expiries(expiries []raftlease.Expired) error {
+	if len(expiries) == 0 {
+		return nil
+	}
+
 	coll, closer := t.mongo.GetCollection(t.collection)
 	defer closer()
 
-	docId := leaseHolderDocId(key.Namespace, key.ModelUUID, key.Lease)
-	t.logger.Infof("expiring lease %q", docId)
+	// Cache all the document idents up front, incase we have to retry the
+	// transaction again. Also this serves as a de-duping process.
+	leaseDocs := make(map[string]leaseDoc)
+	for _, expired := range expiries {
+		key := expired.Key
+		docId := leaseHolderDocId(key.Namespace, key.ModelUUID, key.Lease)
+
+		// We have an existing lease.Key already in the documents that doesn't
+		// have the same holder.
+		if doc, ok := leaseDocs[docId]; ok && doc.Holder != expired.Holder {
+			t.logger.Warningf("multiple expiries with the same key %q, but different holders %q and %q", key, doc.Holder, expired.Holder)
+			continue
+		}
+
+		leaseDocs[docId] = leaseDoc{
+			Key:    key,
+			DocId:  docId,
+			Holder: expired.Holder,
+		}
+	}
+	docIds := set.NewStrings()
+	for _, leaseDoc := range leaseDocs {
+		docIds.Add(leaseDoc.DocId)
+	}
+	sortedDocIds := docIds.SortedValues()
+	t.logger.Infof("expiring leases %v", sortedDocIds)
```

Comment:
Should this be at Info? The existing one was, so we should probably preserve that, but it does make me wonder. I'm very happy that we are doing them as a group (since we do the actual expiration tick as a group), so that you get 1 line rather than 30.